cpp/include/raft/host_buffer.hpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "buffer_base.hpp"
+//TODO: update allocator header
+#include <cuml/common/cuml_allocator.hpp>
+
+namespace raft {
+
+/**
+ * @brief RAII object owning a contigous typed host buffer (aka pinned memory).
+ *        The passed in allocator supports asynchronus allocation and
+ *        deallocation so this can also be used for temporary memory
+ *
+ * @code{.cpp}
+ * template<typename T>
+ * void foo(const raft_handle& h, const T* in_d , T* out_d, ..., cudaStream_t stream ) {
+ *     ...
+ *     host_buffer<T> temp( handle->getHostAllocator(), stream, 0 )
+ *     
+ *     temp.resize(n, stream);
+ *     cudaMemcpyAsync( temp.data(), in_d, temp.size()*sizeof(T), cudaMemcpyDeviceToHost );
+ *     ...
+ *     cudaMemcpyAsync( out_d, temp.data(), temp.size()*sizeof(T), cudaMemcpyHostToDevice );
+ *     temp.release(stream);
+ * }
+ * @endcode
+ */
+template <typename T>
+class host_buffer : public buffer_base<T, hostAllocator> {
+ public:
+  using size_type = typename buffer_base<T, hostAllocator>::size_type;
+  using value_type = typename buffer_base<T, hostAllocator>::value_type;
+  using iterator = typename buffer_base<T, hostAllocator>::iterator;
+  using const_iterator = typename buffer_base<T, hostAllocator>::const_iterator;
+  using reference = typename buffer_base<T, hostAllocator>::reference;
+  using const_reference =
+    typename buffer_base<T, hostAllocator>::const_reference;
+
+  host_buffer() = delete;
+
+  host_buffer(const host_buffer& other) = delete;
+
+  host_buffer& operator=(const host_buffer& other) = delete;
```

Comment:
No support for move constructor and move assignment operator?

And having a constructor (and an assignment operator) taking device_buffer as an input parameter will allow more thrust-like coding style.

e.g.
```
thrust::device_vector<int> d_ints(100, 0);
thrust::host_vector<int> h_ints = d_ints;
```

Reply:
Good point. This will certainly make our code simpler. This should be easy to add.

Reply:
copy operator is hard to do as we expect the caller to also pass host allocator object during ctor. Thus, for now, I have only supported the option of doing copy-ctor from device buffer.